java-connectors/kafka-connect-gcp-common/src/main/java/io/lenses/streamreactor/connect/gcp/common/config/AuthModeSettings.java

```diff
@@ -42,92 +42,109 @@
+   * @return The updated ConfigDef with authentication mode settings defined.
+   */
+  public ConfigDef withAuthModeSettings(ConfigDef configDef) {
+    return configDef
+        .define(
+            authModeKey,
+            Type.STRING,
+            PROP_KEY_DEFAULT,
+            Importance.HIGH,
+            "Authenticate mode, 'credentials', 'file', 'default' or 'none'")
+        .define(
+            credentialsKey,
+            Type.PASSWORD,
+            EMPTY_STRING,
+            Importance.HIGH,
+            "GCP Credentials if using 'credentials' auth mode.")
+        .define(
+            fileKey,
+            Type.STRING,
+            EMPTY_STRING,
+            Importance.HIGH,
+            "File containing GCP Credentials if using 'file' auth mode.  This can be relative from"
+                + " the current working directory of the java process or from the root.  Remember"
+                + " your path format is operating system dependent. (eg for unix-based"
+                + " /home/my/path/file)");
+  }
 
-    /**
-     * Configures the provided ConfigDef with authentication mode settings.
-     *
-     * @param configDef The ConfigDef instance to be updated with authentication mode definitions.
-     * @return The updated ConfigDef with authentication mode settings defined.
-     */
-    public ConfigDef withAuthModeSettings(ConfigDef configDef) {
-        return configDef.define(
-                        authModeKey,
-                        Type.STRING,
-                        PROP_KEY_DEFAULT,
-                        Importance.HIGH,
-                        "Authenticate mode, 'credentials', 'file', 'default' or 'none'"
-                        )
-                .define(
-                        credentialsKey,
-                        Type.PASSWORD,
-                        EMPTY_STRING,
-                        Importance.HIGH,
-                        "GCP Credentials if using 'credentials' auth mode."
-                        )
-                .define(
-                        fileKey,
-                        Type.STRING,
-                        EMPTY_STRING,
-                        Importance.HIGH,
-                        "File containing GCP Credentials if using 'file' auth mode.  This can be relative from the current working directory of the java process or from the root.  Remember your path format is operating system dependent. (eg for unix-based /home/my/path/file)"
-                        );
-    }
+  /**
+   * Parses authentication mode from the provided ConfigMap and returns the corresponding AuthMode instance.
+   *
+   * @param configSource The ConfigSource containing configuration settings.
+   * @return The parsed AuthMode based on the configuration settings.
+   * @throws ConfigException If an invalid or unsupported authentication mode is specified.
+   */
+  public AuthMode parseFromConfig(ConfigSource configSource) {
```

Comment:
In the back of my mind I thought it would be nice to introduce an interface for these to tie them together:

```
  // M is object to materialise from settings
  interface Settings[M] {
      // adds the settings to the config def
     ConfigDef withSettings(ConfigDef configDef);

     // using the config source, materialise the object
     M parseFromConfig(ConfigSource configSource);
   }
```

but that would require standardising on names instead of making each unique.

What do you think?

Reply:
That is actually a very interesting idea, maybe could be better to call it `ConfigDefSetting` but beside that it could actually very nicely tie together many things like propertyname, registering it then making an actual value out of it.

All I'm thinking about is whether any of the basic rules (like SRP and IS) aren't broken (e.g. let's say we create an object out of three properties but there are other places in the code where we're using only two of them to create something else) or if there'll be other scenarios that can break it or it can become less intuitive to implement (e.g. a property is optional, should you return `Optional<M>` then?)

Still, an inspiring idea indeed, I'm just wondering about the consequences down the road 😅